---
title: "Cardiff Pathfinder WS1"
subtitle: "CNV Calling QC Report"
author: 
- "Leon Hubbard"
- "Kimberley Kendall"
- "Elliott Rees"
- "George Kirov"
date: "`r format(Sys.Date(), '%d-%m-%Y')`"

output: 
  prettydoc::html_pretty:
    theme: leonids
    highlight: vignette
    fig_caption: yes
    toc: FALSE
    toc_depth: 2
    number_sections: FALSE
params:
  DATASET: ""
---

```{r Required Packages,include=F}
library("ggplot2")
library("gridExtra")
library("tidyverse") # Package version: 1.2.1
library("ggthemes") # Package version: 4.0.1
library("scales") # Package version 1.0.0
library("RColorBrewer") # Package version 1.1-2
library("evaluate") # devtools::install_github('hadley/evaluate')
library("data.table") # Package version 1.12.0
library("rvest") # Package version 0.3.2
library("data.table")
library("limma")
```

```{r setup, include=F}
library("knitr")
library("prettydoc")
knitr::opts_chunk$set(echo=F, tidy=T, cache=F, message=F, error=F, warning=F, comment=NA, out.width = '95%')
opts_knit$set(width = 75)
options(formatR.arrow=T, max.print = "75")
```

```{r Define variables and paths,include=F}
evaluate::inject_funs(system = function(...) cat(base::system(..., intern = TRUE), sep = "\n"))

miscfolder="~/Pathfinder/CNVCallingPipeExtra/"
DATASET=""
PENNCNV="~/Pathfinder/CNVCallingPipeExtra/PennCNV-1.0.5/"
PERL="~/Pathfinder/CNVCallingPipeExtra/perl-5.14.2/perl"
outpath="/scratch/c.c1002680"
#outpath="~/610_550/"
outpath=paste(outpath,"/",DATASET,"/",sep="")
if(dir.exists("NeuroDevelopmentalPlots")==F){
dir.create("NeuroDevelopmentalPlots")
}
```

```{r CNV Calling,include=T}
### We use kcolumn.pl to write files for each individual (b-allele freq,log R rations)
### - heading refers to the number of trailing columns before the sample information starts
### -split refers to the number of columns for each individual (usually gtype, ballele freq and LRR)

system(paste(PERL," ",PENNCNV,"kcolumn.pl ",DATASET,".txt split 3 -heading 3 -tab --name_by_header --out ",outpath,DATASET,sep=""))
system(paste("ls | grep ",DATASET," | grep -v txt | grep -v out | grep -v err | grep -v Rmd > ",outpath,"PennCNV.list",sep=""))
### Generate CNV calls
system(paste(PERL," ",PENNCNV,"detect_cnv.pl -test -hmm  ",miscfolder,"hhall.hmm -listfile PennCNV.list -pfb ",miscfolder,"psychChip_2017.pfb --confidence --log ",outpath,DATASET,".Calls.log --out ",outpath,DATASET,".Calls -gc ",miscfolder,"psychChip_2017.gcmodel",sep=""))

### Merge CNVs 
system(paste(PERL," ",PENNCNV,"clean_cnv.pl combineseg ",DATASET,".Calls -fraction 0.5 -signalfile ",miscfolder,"psychChip_2017.pfb --output ",outpath,DATASET,".Calls.Merged",sep=""))

# Generate CNV calls 
system(paste(PERL," ",PENNCNV,"filter_cnv.pl -length -numsnp 10 ",outpath,DATASET,".Calls.Merged -qclogfile ",outpath,DATASET,".Calls.log -qcpassout ",outpath,DATASET,".qcpass -qcsumout ",DATASET,".qcsum -out ",outpath,DATASET,".goodcnv",sep=""))

cnv.qual=read.table(file=paste(outpath,DATASET,".qcsum",sep=""),header=T,stringsAsFactors = F)
cnv.include=read.table(paste(outpath,DATASET,".qcpass",sep=""),header=F)

cnv.flag.fail=cnv.qual[which(cnv.qual$GROUP != "PennCNV Pass"),]
```
# Overview  
As part of the MRC Mental Health Data Pathfinder grant, we are calling CNVs to identify individuals who may carry a known neurodevelopmental risk CNV.  
This pipeline is designed to perform the following steps:  

**1)** Perform CNV calling using [PennCNV(v1.04)](http://penncnv.openbioinformatics.org/en/latest/)  
**2)** Perform quality control and remove individuals/cnvs that fail to meet specific parameters  
**3)** Cross-reference remaining high quality CNV calls against a list of known neurodevelopmental CNVs (see [Kendall el al, 2019](https://jamanetwork.com/journals/jamapsychiatry/fullarticle/2730725))  
**4)** Identify whether these CNVs pass criteria specific to each locus to be reliably called as being ND CNVs.

## CNV QC Parameters  
CNV calling was performed using PennCNV for the `r DATASET` sample using B-allele frequencies and LogR ratios exported from Illumina Genome Studio. The PFB file used for GC correction was CNVs were called in `r nrow(cnv.qual)` individuals.  

Poor performing samples can be identified using a variety of QC metrics from PennCNV:  
**Log R Ratios (LRR)**: log2 of R (observed SNP intensity) / expected R intensity. CNV deletions and duplications have decreased and increased LRR, respectively.  
**B-allele frequencies (BAF)**: are a normalised measure of the allelic intensity ratio. The expected BAF for AA, AB and BB genotypes is 0, 0.5 and 1, respectively. SNPs in deleted segments of DNA should all have homozygous genotypes, therefore they should all have a BAF of ~0 or ~1. The BAF of heterozygous SNPs in a duplicated DNA segment is either ~0.33 for the genotype AAB or ~0.66 for the genotype ABB, depending on which allele is affected.   
**Wavefactor**: patterns of signal intensities across all chromosomes, where different samples may show highly variable magnitude of waviness.  
**Number of CNVs per individual**: if individuals carry a large number of CNVs this is indicative of a poor quality sample. Note, these individuals may still pass QC parameters for SNP genotype QC.

We can visualise different QC metrics to visually identify outliers, indicative of poor performing samples.   
**1)** Wavefactor vs Number of CNVs  
**2)** Log R Ratio vs Number of CNVs  
**2)** B-allele frequency drift  
**3)** Standard deviation of Log R Ratios  

```{r CNV Plots,include=T,fig.align='center'}
cnv.include$GROUP="PennCNV Pass"
cnv.qual=merge(cnv.qual,cnv.include,by.x="File",by.y="V1",sort=F,all.x=T)
cnv.qual[is.na(cnv.qual)] <- "PennCNV Fail"
plot1=ggplot(cnv.qual,aes(x=WF,y=NumCNV,color=GROUP))+
  geom_point(shape=1)+
  xlab("Wavefactor")+
  ylab("N CNVs")+
  scale_color_manual(values=c("#E69F00","#56B4E9"))+
  theme(legend.position="none")

plot2=ggplot(cnv.qual,aes(x=LRR_SD,y=NumCNV,color=GROUP))+
  geom_point(shape=1)+
  xlab("Mean LRR")+
  ylab("N CNVs")+
  scale_color_manual(values=c("#E69F00","#56B4E9"))+
  theme(legend.position="bottom")


plot3=ggplot(cnv.qual, aes(x=BAF_drift)) + geom_histogram(alpha=.5, position="identity",colour="black", fill="white")+
  xlab("B-Allele Frequency Drift")+
  ylab("Frequency")

plot4=ggplot(cnv.qual, aes(x=LRR_SD)) + geom_histogram(colour="black", fill="white")+
  xlab("LRR SD")+
  ylab("Frequency")
no.outliers.removed=grid.arrange(arrangeGrob(plot1,plot2,ncol=1),arrangeGrob(plot3,plot4,ncol=1),ncol=2, widths=c(1,1))

```

### Criteria for identifying sample outliers  
**1)** Their total number of CNVs >100  
**2)** LogR Ratio standard deviation >0.2   
**3)** Wavefactor is < -0.03 or > 0.03  
**4)** Samples have low genotyping rates (<0.95)

### Venn Diagrams - Identifying Outliers

The numbers within each circle represent the number of people excluded at the specified threshold.

```{r CNV Plots venndia,include=T,fig.align='center'}
 

LRR_SD.range=0.2
NCNV.range=100
WF.range=0.03

LRR_SD.vd=cnv.qual$LRR_SD>=as.numeric(LRR_SD.range[1])
NCNV.vd=cnv.qual$NumCNV>=as.numeric(NCNV.range[1])
WF.vd=(cnv.qual$WF<=(-1*as.numeric(WF.range[1])) |  cnv.qual$WF>= as.numeric(WF.range[1]))
vd.all=cbind(LRR_SD.vd,NCNV.vd,WF.vd)
venn.diagram.lowest.thres=vennCounts(vd.all)
vennDiagram(vd.all,names = c(paste("LRR_SD"),paste("NCNV"),paste("WF")))
title(paste("Individuals excluded using the following parameters: \nLRR>",LRR_SD.range[1]," ; NCNVs>",NCNV.range[1]," ; WF>",WF.range[1],sep=""))
```

After excluding individuals that do not pass this QC:

```{r outliers removed,align='center'}

exclude.individuals=cnv.qual[which(as.numeric(as.character(cnv.qual$NumCNV)) >= 100 | as.numeric(as.character(cnv.qual$WF)) >=0.03 | as.numeric(as.character(cnv.qual$WF))<=-0.03 | as.numeric(as.character(cnv.qual$LRR_SD)) >=0.2),]


cnv.qual=cnv.qual[! cnv.qual$File %in% unlist(exclude.individuals$File),]
plot1no=ggplot(cnv.qual,aes(x=WF,y=NumCNV,color=GROUP))+
  geom_point(shape=1)+
  xlab("Wavefactor")+
  ylab("N CNVs")+
  scale_color_manual(values=c("#56B4E9"))+
  theme(legend.position="none")

plot2no=ggplot(cnv.qual,aes(x=LRR_SD,y=NumCNV,color=GROUP))+
  geom_point(shape=1)+
  xlab("Mean LRR")+
  ylab("N CNVs")+
  scale_color_manual(values=c("#56B4E9"))+
  theme(legend.position="bottom")

plot3no=ggplot(cnv.qual, aes(x=BAF_drift)) + geom_histogram(alpha=.5, position="identity",colour="black", fill="white")+
  xlab("B-Allele Frequency Drift")+
  ylab("Frequency")
  #geom_vline(xintercept=sd(cnv.qual[which(cnv.qual$GROUP=="PennCNV Pass"),]$BAF_drift)*2, linetype="dashed", color = "Red")


plot4no=ggplot(cnv.qual, aes(x=LRR_SD)) + geom_histogram(colour="black", fill="white")+
  xlab("LRR SD")+
  ylab("Frequency")

grid.arrange(arrangeGrob(plot1no,plot2no,ncol=1),arrangeGrob(plot3no,plot4no,ncol=1),ncol=2, widths=c(1,1))
```

### Sample /CNV exclusions
`r nrow(exclude.individuals)` individuals failed QC (either > 100 CNVs, WF>0.03 or < -0.03 or LRR SD > 0.2).

```{r high quality individuals}
#high.qual=cnv.qual[pass.individual.qc,]

CNV.Calls=as.data.frame(fread(input=paste(outpath,DATASET,".goodcnv",sep=""),header=F,stringsAsFactors = F))
names(CNV.Calls)=c("COORDS","NPROBES","SIZE","TYPE","ID","START_PROBE","END_PROBE","CONF")
CNV.Calls$NPROBES=gsub("numsnp=","",CNV.Calls$NPROBES)
CNV.Calls$SIZE=gsub("length=","",CNV.Calls$SIZE)
CNV.Calls$SIZE=gsub(",","",CNV.Calls$SIZE)
CNV.Calls$CHR=gsub("chr","",do.call(rbind,strsplit(CNV.Calls$COORDS,split=":"))[,1])
CNV.Calls$CONF=gsub("conf=","",CNV.Calls$CONF)
CNV.Calls$CONF=as.numeric(as.character(CNV.Calls$CONF))
CNV.Calls$START=as.numeric(gsub(",","",do.call(rbind,strsplit(gsub("chr","",do.call(rbind,strsplit(CNV.Calls$COORDS,split="-"))[,1]),split=":"))[,2]))
CNV.Calls$END=as.numeric(gsub(",","",do.call(rbind,strsplit(CNV.Calls$COORDS,split="-"))[,2]))
CNV.Calls=CNV.Calls[which(as.numeric(as.character(CNV.Calls$SIZE))>100000) & as.numeric(as.character(CNV.Calls$NPROBES))>=20 & CNV.Calls$CONF>=10,]
#CNV.Calls.Exclude=CNV.Calls[which(as.numeric(as.character(CNV.Calls$SIZE))<20000 | as.numeric(as.character(CNV.Calls$NPROBES))<10 | CNV.Calls$CONF<10),]
CNV.Calls.Exclude=CNV.Calls[which(as.numeric(as.character(CNV.Calls$SIZE))<100000 | as.numeric(as.character(CNV.Calls$NPROBES))<10 | CNV.Calls$CONF<10),]

#CNV.Calls=CNV.Calls[grep("_R1$",CNV.Calls$ID),]

CNV.Calls.HQ=CNV.Calls[!CNV.Calls$ID %in% unlist(exclude.individuals$File),]

hundredKB=CNV.Calls[which(as.numeric(CNV.Calls.HQ$SIZE) > 100000),]
hundredKB=CNV.Calls
```

```{r parse input file}

cnv.neurodev=read.table(file=paste(miscfolder,"CNVS.KK.2019.Sorted.txt",sep=""),sep="\t",stringsAsFactors=F)
cnv.neurodev$CHR=gsub("chr","",do.call(rbind,strsplit(cnv.neurodev$V3,split=":"))[,1])
cnv.neurodev$START=as.numeric(gsub(",","",do.call(rbind,strsplit(gsub("chr","",do.call(rbind,strsplit(cnv.neurodev$V3,split="-"))[,1]),split=":"))[,2]))
cnv.neurodev$END=as.numeric(gsub(",","",do.call(rbind,strsplit(cnv.neurodev$V3,split="-"))[,2]))
cnv.neurodev$CLASS=do.call(rbind,lapply(1:nrow(cnv.neurodev),function(x){
    ifelse(length(grep("del",cnv.neurodev[x,1]))>0,yes=1,no=3)
}))
names(cnv.neurodev)[2]=c("LOCUS")
cnv.neurodev=cnv.neurodev[,c(1,4:7)]
```

```{r get neurodev cnvs,warning=FALSE}
#CNV.Calls=CNV.Calls.HQ[which(as.numeric(as.character(CNV.Calls.HQ$SIZE))>=50000),]
#CNV.Calls=CNV.Calls.HQ[which(as.numeric(as.character(CNV.Calls.HQ$SIZE))>=50000),]

cnv.neuro.beta=unique(as.data.frame(do.call(rbind,lapply(1:nrow(CNV.Calls),function(x){
  cnv.chr=CNV.Calls[x,9]
  cnv.start=CNV.Calls[x,10]
  cnv.end=CNV.Calls[x,11]

### CNV spans entire region
  cnv.neurodev.match=cnv.neurodev[which(cnv.neurodev$CHR==cnv.chr & cnv.start<=cnv.neurodev$START & cnv.end>=cnv.neurodev$END),] 
  
   if(nrow(cnv.neurodev.match)>0){
     cnv.prop=1
     cnv.neurodev.match=cbind(cnv.neurodev.match,cnv.prop)
   } else if(nrow(cnv.neurodev.match)==0){
    
    ### CNV start is less than the start of the nd cnv and end of the CNV is less than than end of the nd cnv
    cnv.neurodev.match=cnv.neurodev[which(cnv.neurodev$CHR==cnv.chr & cnv.start<=cnv.neurodev$START & cnv.end<=cnv.neurodev$END & cnv.end>=cnv.neurodev$START),]
    
    if(nrow(cnv.neurodev.match)>0){
      cnv.neurodev.match.start=cnv.neurodev.match$START
      cnv.neurodev.match.end=cnv.neurodev.match$END
      
      cnv.prop=(cnv.end-cnv.neurodev.match$START)/(cnv.neurodev.match.end-cnv.neurodev.match.start)
      cnv.neurodev.match=cbind(cnv.neurodev.match,cnv.prop)
      ### CNV start is greater than start of nd cnv and ends after the end of the nd region
    } else if(nrow(cnv.neurodev.match)==0){
        cnv.neurodev.match=cnv.neurodev[which(cnv.neurodev$CHR==cnv.chr & cnv.start>=cnv.neurodev$START & cnv.end>=cnv.neurodev$END & cnv.start<=cnv.neurodev$END),]
    
        if(nrow(cnv.neurodev.match)>0){
      cnv.neurodev.match.start=cnv.neurodev.match$START
      cnv.neurodev.match.end=cnv.neurodev.match$END
      
      cnv.prop=(cnv.neurodev.match$END-cnv.start)/(cnv.neurodev.match.end-cnv.neurodev.match.start)
      cnv.neurodev.match=cbind(cnv.neurodev.match,cnv.prop)
      
      } else if(nrow(cnv.neurodev.match)==0){
        ### CNV falls inside nd region
        cnv.neurodev.match=cnv.neurodev[which(cnv.neurodev$CHR==cnv.chr & cnv.start>=cnv.neurodev$START & cnv.end<=cnv.neurodev$END),]
  
      if(nrow(cnv.neurodev.match)>0){
      cnv.neurodev.match.start=cnv.neurodev.match$START
      cnv.neurodev.match.end=cnv.neurodev.match$END
      
      cnv.prop=(cnv.end-cnv.start)/(cnv.neurodev.match.end-cnv.neurodev.match.start)
      cnv.neurodev.match=cbind(cnv.neurodev.match,cnv.prop)
      }}}}
  
  if(nrow(cnv.neurodev.match)>0){
    cbind(CNV.Calls[x,],cnv.neurodev.match)
  }
  
}))))

cnv.neuro.beta$TYPE=do.call(rbind,strsplit(cnv.neuro.beta$TYPE,split="="))[,2]
cnv.neuro.beta=cnv.neuro.beta[which(cnv.neuro.beta$TYPE==cnv.neuro.beta$CLASS),]

cnv.neuro.beta.fifpc=cnv.neuro.beta

#geneloc=read.table(file=paste("~/Hybrid.Magma/Data/NCBI37.3.gene.loc",sep=""),header=F,stringsAsFactors = F)
geneloc=read.table(file=paste(miscfolder,"NCBI37.3.gene.loc",sep=""),header=F,stringsAsFactors = F)

cnv.neuro.beta.genes=as.data.frame(do.call(rbind,lapply(1:nrow(cnv.neuro.beta.fifpc),function(x){
  a=cnv.neuro.beta.fifpc[x,]
  cnv.chr=a[1,9]
  cnv.start=a[1,10]
  cnv.end=a[1,11]
  
  geneloc.chr=geneloc[which(geneloc$V2==cnv.chr),]
  
  ### CNV covers whole gene
  gene.whole=geneloc.chr[which(cnv.start<geneloc.chr$V3 & cnv.end>geneloc.chr$V4),]
  gene.start=geneloc.chr[which(cnv.start<geneloc.chr$V3 & cnv.end<geneloc.chr$V4 & cnv.end>geneloc.chr$V3),]
  gene.end=geneloc.chr[which(cnv.start>geneloc.chr$V3 & cnv.end>geneloc.chr$V4 & cnv.start<geneloc.chr$V4),]
  gene.mid=geneloc.chr[which(cnv.start>geneloc.chr$V3 & cnv.end<geneloc.chr$V4),]

  
  if(nrow(gene.whole)>0 | nrow(gene.start)>0 | nrow(gene.end)>0 | nrow(gene.mid)>0){
  a$GENES=paste(unique(c(gene.whole$V6,gene.start$V6,gene.end$V6,gene.mid$V6)),collapse="|")
  a
  } else{
    a$GENES=""
    a
  }
})))

### Check for duplicate IDs. It would be unusual for an individual to have two different ND CNVs
### Most likely cause for dupliucates are having a smaller nested CNV within a larger one, where both are ND 
### Loop over the IDs, check to see if the CNV coordinates are duplicates, and if so just report the largest CNV. 

#if(nrow(cnv.neuro.beta.genes[which(duplicated(cnv.neuro.beta.genes$ID)),])>0){

#nested.cnv.dups=as.data.frame(do.call(rbind,lapply(cnv.neuro.beta.genes[which(duplicated(cnv.neuro.beta.genes$ID)),]$ID,function(x){
#  cnv.dups=cnv.neuro.beta.genes[which(cnv.neuro.beta.genes$ID==x),]
  
#  if(length(unique(cnv.dups$COORDS))==1){
    
    ### Get the smaller of the two CNVs (the smaller one will be the nested one)
#    cnv.nested=which.min(cnv.dups[,15]-cnv.dups[,14])
#    cnv.exclude=cnv.dups[cnv.nested,c(5,12)]
#  }
#})))
#}

#if(exists("nested.cnv.dups")==T){
#cnv.neuro.beta.fifpc=cnv.neuro.beta.genes[!(cnv.neuro.beta.genes$ID %in% unlist(nested.cnv.dups[,1]) & cnv.neuro.beta.genes$V1 %in% unlist(nested.cnv.dups[,2])),]
#}

#miscfolder="~/Downloads/"
#### NRXN1 exon processing
nrxn1=read.table(file=paste(miscfolder,"exon.NCBI.NRXN1",sep=""),header=T,stringsAsFactors = F)
nrx1.exon.start=unlist(strsplit(nrxn1$exonStarts,split = ","))
nrx1.exon.end=unlist(strsplit(nrxn1$exonEnds,split = ","))
nrx1.exon.coords=unique(as.data.frame(do.call(rbind,lapply(1:length(nrx1.exon.start),function(x){
  nrx1.exon.start.loop=nrx1.exon.start[x]
  nrx1.exon.end.loop=nrx1.exon.end[x]
  cbind(nrx1.exon.start.loop,nrx1.exon.end.loop)
}))))
nrx1.exon.coords$EXON=paste("NRXN1 Exon ",seq(1:nrow(nrx1.exon.coords)),sep="")
###

### YWHAE exon processing
YWHAE=read.table(file=paste(miscfolder,"exon.YWHAE",sep=""),header=T,stringsAsFactors = F,fill = T)
YWHAE.exon.start=unlist(strsplit(YWHAE$exonStarts,split = ","))
YWHAE.exon.end=unlist(strsplit(YWHAE$exonEnds,split = ","))
YWHAE.exon.coords=unique(as.data.frame(do.call(rbind,lapply(1:length(YWHAE.exon.start),function(x){
  YWHAE.exon.start.loop=YWHAE.exon.start[x]
  YWHAE.exon.end.loop=YWHAE.exon.end[x]
  cbind(YWHAE.exon.start.loop,YWHAE.exon.end.loop)
}))))
YWHAE.exon.coords$EXON=paste("YWHAE Exon ",seq(1:nrow(YWHAE.exon.coords)),sep="")
###

### PAFAH1B1 exon processing
PAFAH1B1=read.table(file=paste(miscfolder,"exon.PAFAH1B1",sep=""),header=T,stringsAsFactors = F,fill = T)
PAFAH1B1.exon.start=unlist(strsplit(PAFAH1B1$exonStarts,split = ","))
PAFAH1B1.exon.end=unlist(strsplit(PAFAH1B1$exonEnds,split = ","))
PAFAH1B1.exon.coords=unique(as.data.frame(do.call(rbind,lapply(1:length(PAFAH1B1.exon.start),function(x){
  PAFAH1B1.exon.start.loop=PAFAH1B1.exon.start[x]
  PAFAH1B1.exon.end.loop=PAFAH1B1.exon.end[x]
  cbind(PAFAH1B1.exon.start.loop,PAFAH1B1.exon.end.loop)
}))))
PAFAH1B1.exon.coords$EXON=paste("PAFAH1B1 Exon ",seq(1:nrow(PAFAH1B1.exon.coords)),sep="")
###

cnv.neuro.beta.genes$CRITERIA_MET=0
cnv.neuro.beta.genes$SIZE=as.numeric(cnv.neuro.beta.genes$SIZE)

cnvs.unique=unique(cnv.neuro.beta.genes$V1)

cnv.patho.criteria=as.data.frame(do.call(rbind,lapply(cnvs.unique,function(x){
#cnv.patho.criteria[1,12]

  CNV.ND=cnv.neuro.beta.genes[which(cnv.neuro.beta.genes$V1==x),]
  
  as.data.frame(do.call(rbind,lapply(1:nrow(CNV.ND),function(y){
   CNV.ND.Line=CNV.ND[y,] 
   
        # 1q21.1	del/dup		 Size	>50%	of	critical	region
     if((x=="1q21.1 del" | x=="1q21.1 dup") & CNV.ND.Line$cnv.prop>0.5){
      CNV.ND.Line$CRITERIA_MET=1    
        } 
  
     # 1p36 del/dup		 Size	>50%	of	critical	region,	affecting	GABRD
     if((x=="1p36 del (GABRD)" | x=="1p36 dup (GABRD)") & length(grep("GABRD",CNV.ND.Line$GENES))>0 & CNV.ND.Line$cnv.prop>0.5){
      CNV.ND.Line$CRITERIA_MET=1    
        } 
  
      # TAR del/dup		 Size	>50%	of	critical	region
        if((x=="TAR del" | x=="TAR dup") & CNV.ND.Line$cnv.prop>0.5){
      CNV.ND.Line$CRITERIA_MET=1    
        } 
   
    # NRXN1 del, must include complete deletion of at least one exon
        if((x=="NRXN1 del") & length(grep("NRXN1",CNV.ND.Line$GENES))>0){
          
          
          nrxn1.exon.count=as.data.frame(do.call(rbind,lapply(1:nrow(nrx1.exon.coords),function(z){
            CNV.ND.Line.Start=CNV.ND.Line[1,10]
            CNV.ND.Line.End=CNV.ND.Line[1,11]

            exon.whole=nrx1.exon.coords[which(as.numeric(as.character(nrx1.exon.coords[z,1]))>CNV.ND.Line.Start & as.numeric(as.character(nrx1.exon.coords[z,2]))<CNV.ND.Line.End),]
            
            if(nrow(exon.whole)>0){
              exon.whole
            }
            
          })))
          
          if(nrow(nrxn1.exon.count)>0){
      CNV.ND.Line$CRITERIA_MET=1    
          }
        }
   
      # 2q11.2	del/dup	Size	>50%	of	critical	region,	affecting	both LMAN2L and	ARID5A
     if((x=="2q11.2 del" | x=="2q11.2 dup") & length(grep("LMAN2L",CNV.ND.Line$GENES))>0 & length(grep("ARID5A",CNV.ND.Line$GENES))>0 & CNV.ND.Line$cnv.prop>0.5){
      CNV.ND.Line$CRITERIA_MET=1    
        } 
   
   # 2q13	del/dup		 Size	>50%	of	critical	region		
        if((x=="2q13 del" | x=="2q13 dup") & CNV.ND.Line$cnv.prop>0.5){
      CNV.ND.Line$CRITERIA_MET=1    
        } 
   
  # 2q13	del/dup	(NPHP1)		 Size	>50%	of	critical	region,	affecting	NPHP1
           if((x=="2q13 del (NPHP1)" | x=="2q13 dup (NPHP1)") & length(grep("NPHP1",CNV.ND.Line$GENES))>0 & CNV.ND.Line$cnv.prop>0.5){
      CNV.ND.Line$CRITERIA_MET=1    
        } 
   
   # 2q21.1	del/dup		 Size	>50%	of	critical	region		
           if((x=="2q21.1 del" | x=="2q21.1 dup") & CNV.ND.Line$cnv.prop>0.5){
      CNV.ND.Line$CRITERIA_MET=1    
        } 
   
   # 2q37	del	(HDAC4)		 Size	>50%	of	critical	region,	affecting	HDAC4
           if((x=="2q37 del (HDAC4)" | x=="2q37 dup (HDAC4)") & CNV.ND.Line$cnv.prop>0.5 & length(grep("HDAC4",CNV.ND.Line$GENES))>0){
      CNV.ND.Line$CRITERIA_MET=1    
        } 
   
   # 3q29	del/dup		 Size	>50%	of	critical	region		
           if((x=="3q29	del" | x=="3q29	dup") & CNV.ND.Line$cnv.prop>0.5){
      CNV.ND.Line$CRITERIA_MET=1    
           } 
   
   # Wolf‐Hirschhorn	del/dup		 Size	>50%	of	critical	region		
           if((x=="Wolf‐Hirschhor del" | x=="Wolf‐Hirschhor dup") & CNV.ND.Line$cnv.prop>0.5){
      CNV.ND.Line$CRITERIA_MET=1    
           } 
   
# Sotos	Syn/5q35	dup		 Size	>50%	of	critical	region		
           if(x=="Sotos Syn/5q35 dup" & CNV.ND.Line$cnv.prop>0.5){
      CNV.ND.Line$CRITERIA_MET=1    
           } 
   
 # Williams	Beuren	Syn	del/dup	Size	>50%	of	critical	region		
               if((x=="Williams	Beuren Syn del" | x=="Williams Beuren Syn dup") & CNV.ND.Line$cnv.prop>0.5){
      CNV.ND.Line$CRITERIA_MET=1    
               } 
   
   # 8p23.1	del/dup		 At	least	1Mbp	of	critical	region		
                   if((x=="8p23.1 del" | x=="8p23.1 dup") & CNV.ND.Line$SIZE>1000000){
      CNV.ND.Line$CRITERIA_MET=1    
               }
   
   # 9q34	del/dup	(EHMT1)		 At	least	1Mbp	CNVs,	including	EHMT1
      if(x=="9q34 del (EHMT1)" & CNV.ND.Line$SIZE>1000000 & length(grep("EHMT1",CNV.ND.Line$GENES))>0){
      CNV.ND.Line$CRITERIA_MET=1    
               }
   
  # 10q11.21q11.23	del/dup		 Size	>50%	of	critical	region		
      if((x=="10q11.21q11.23 del" | x=="10q11.21q11.23 dup") & CNV.ND.Line$cnv.prop>0.5){
      CNV.ND.Line$CRITERIA_MET=1    
               }
   
   # 10q23	del/dup		 At	least	1Mbp,	including	NRG3 and	GRID1
      if((x=="10q23 del" | x=="10q23 dup") & CNV.ND.Line$SIZE>1000000 & length(grep("NRG3",CNV.ND.Line$GENES))>0 & length(grep("GRID1",CNV.ND.Line$GENES))>0){
      CNV.ND.Line$CRITERIA_MET=1    
               }
   
      # Potocki‐Shaffer	Syn	del/11p11.2	dup (EXT2), size	>50%	of	critical	region,	including	EXT2

         if(x=="Potocki-Shaffer syndrome del (EXT2)" & CNV.ND.Line$cnv.prop>0.5 & length(grep("EXT2",CNV.ND.Line$GENES))){
      CNV.ND.Line$CRITERIA_MET=1    
               }
   
   # 15q11.2	del/dup		 Size	>50%	of	critical	region		
            if((x=="15q11.2 del" | x=="15q11.2 dup") & CNV.ND.Line$cnv.prop>0.5){
      CNV.ND.Line$CRITERIA_MET=1    
               }
   
      # 15q11.2 del/dup BP1-BP2		 Size	>50%	of	critical	region		
            if((x=="15q11.2 del BP1-BP2" | x=="15q11.2 dup BP1-BP2") & CNV.ND.Line$cnv.prop>0.5){
      CNV.ND.Line$CRITERIA_MET=1    
               }
   
   # PWS	del/dup		 Full	critical	region,	~4Mbp	
         if((x=="(PWS/AS del" | x=="(PWS/AS dup") & CNV.ND.Line$SIZE>4000000 & CNV.ND.Line$cnv.prop>0.8){
      CNV.ND.Line$CRITERIA_MET=1    
               }
   
      # 15q13.3 del BP4-BP5		 Size	>50%	of	critical	region
      if(x=="15q13.3 del BP4-BP5" & CNV.ND.Line$cnv.prop>0.5){
      CNV.ND.Line$CRITERIA_MET=1    
      }
   
# 15q24	del/dup		 At	least	1Mbp	between	the	A‐E	intervals	
         if((x=="15q24 del" | x=="15q24 dup") & CNV.ND.Line$SIZE>1000000){
      CNV.ND.Line$CRITERIA_MET=1    
               }
   
   
   
# 15q25	del/dup		 At	least	1Mbp	between	the	A‐D	intervals		
         if(x=="15q25 del" & CNV.ND.Line$SIZE>1000000){
      CNV.ND.Line$CRITERIA_MET=1    
               }
   
   # 16p13.11	del/dup		 Size	>50%	of	critical	region
      if((x=="16p13.11 del" | x=="16p13.11 dup") & CNV.ND.Line$cnv.prop>0.5){
      CNV.ND.Line$CRITERIA_MET=1    
               }
   
# 16p12.1	del		 Size	>50%	of	critical	region
         if(x=="16p12.1 del" & CNV.ND.Line$cnv.prop>0.5){
      CNV.ND.Line$CRITERIA_MET=1    
               }
   
   

   # 16p11.2	distal	del/distal	dup		 Size	>50%	of	critical	region		
            if((x=="16p11.2 distal del" | x=="16p11.2 distal dup") & CNV.ND.Line$cnv.prop>0.5){
      CNV.ND.Line$CRITERIA_MET=1    
               }
   
   # 16p11.2	del/dup		 Size	>50%	of	critical	region
            if((x=="16p11.2 del" | x=="16p11.2 dup") & CNV.ND.Line$cnv.prop>0.5){
      CNV.ND.Line$CRITERIA_MET=1    
               }
   
   
   
# 17p13.3	del/dup	(YWHAE)		 Exonic	deletions;	whole	gene	duplications
            
        if((x=="17p13.3 del (YWHAE)" | x=="17p13.3	dup (YWHAE") & CNV.ND.Line$cnv.prop>0.5 & length(grep("YWHAE",CNV.ND.Line$GENES))>0){
                 
            YWHAE.exon.count=as.data.frame(do.call(rbind,lapply(1:nrow(YWHAE.exon.coords),function(z){
            CNV.ND.Line.Start=CNV.ND.Line[1,10]
            CNV.ND.Line.End=CNV.ND.Line[1,11]

            exon.whole=YWHAE.exon.coords[which(as.numeric(as.character(YWHAE.exon.coords[z,1]))>CNV.ND.Line.Start & as.numeric(as.character(YWHAE.exon.coords[z,2]))<CNV.ND.Line.End),]
            
            if(nrow(exon.whole)>0){
              exon.whole
            }
            
          })))
          
          if(nrow(YWHAE.exon.count)>0){
      CNV.ND.Line$CRITERIA_MET=1    
          }
        }
                 
  
# 17p13.3	del/dup	(PAFAH1B1) Exonic	deletions;	whole	gene	duplications
                if((x=="17p13.3 del (PAFAH1B1)" | x=="17p13.3	dup (PAFAH1B1") & CNV.ND.Line$cnv.prop>0.5 & length(grep("PAFAH1B1",CNV.ND.Line$GENES))>0){
            PAFAH1B1.exon.count=as.data.frame(do.call(rbind,lapply(1:nrow(PAFAH1B1.exon.coords),function(z){
            CNV.ND.Line.Start=CNV.ND.Line[1,10]
            CNV.ND.Line.End=CNV.ND.Line[1,11]

            exon.whole=PAFAH1B1.exon.coords[which(as.numeric(as.character(PAFAH1B1.exon.coords[z,1]))>CNV.ND.Line.Start & as.numeric(as.character(PAFAH1B1.exon.coords[z,2]))<CNV.ND.Line.End),]
            
            if(nrow(exon.whole)>0){
              exon.whole
            }
            
          })))
          
          if(nrow(PAFAH1B1.exon.count)>0){
      CNV.ND.Line$CRITERIA_MET=1    
          }
                }
# 17q11.2	del/dup	(NF1)	>50%	of	critical	region,	affecting	NF1
            if((x=="17q11.2 del (NF1)" | x=="17q11.2 dup (NF1)") & CNV.ND.Line$cnv.prop>0.5 & length(grep("NF1",CNV.ND.Line$GENES))>0){
      CNV.ND.Line$CRITERIA_MET=1    
               }
   
   
# Smith‐Magenis/Potocki‐Lupski	Syndrome	Size	>50%	of	critical	region		

            if((x=="Smith-Magenis syndrome del" | x=="Potocki-Lupski syndrome dup") & CNV.ND.Line$cnv.prop>0.5){
      CNV.ND.Line$CRITERIA_MET=1    
            }
   
   # 17q11.2	del/dup	(NF1)		 Size	>50%	of	critical	region,	affecting	NF1
         if((x=="17q11.2 del (NF1)" | x=="17q11.2 dup (NF1)") & CNV.ND.Line$cnv.prop>0.5 & length(grep("NF1",CNV.ND.Line$GENES))>0){
      CNV.ND.Line$CRITERIA_MET=1    
               }
   
   # 17q12	del/dup		 Size	>50%	of	critical	region		
         if((x=="17q12 del" | x=="17q12 dup" | x=="Renal cysts and diabetes syndrome del") & CNV.ND.Line$cnv.prop>0.5){
      CNV.ND.Line$CRITERIA_MET=1    
               }
   
   # 17q21.31	del/dup		 Size	>50%	of	critical	region		
         if((x=="17q21.31 del" | x=="17q21.31 dup") & CNV.ND.Line$cnv.prop>0.5){
      CNV.ND.Line$CRITERIA_MET=1    
               }
   
   # 17q23.1q23.2	del		 Size	>50%	of	critical	region		
         if(x=="17q23.1q23.2 del" & CNV.ND.Line$cnv.prop>0.5){
      CNV.ND.Line$CRITERIA_MET=1    
               }

   # 17p12       del              Size   >50%    of      critical        region
         if(x=="17p12 del" & CNV.ND.Line$cnv.prop>0.5){
      CNV.ND.Line$CRITERIA_MET=1
               }
   
   # 22q11.2	del/dup		 Size	>50%	of	critical	region		
         if((x=="22q11.2 del" | x=="22q11.2 dup") & CNV.ND.Line$cnv.prop>0.5){
      CNV.ND.Line$CRITERIA_MET=1    
               }
   
   # 22q11.2	distal	del/dup		 Size	>50%	of	critical	region
         if((x=="22q11.2 distal del" | x=="22q11.2 distal dup") & CNV.ND.Line$cnv.prop>0.5){
      CNV.ND.Line$CRITERIA_MET=1    
         }

      # SHANK3 del/dup	 At	least	1Mbp	CNVs,	including	SHANK3
     if((x=="SHANK3 del" | x=="SHANK3 dup") & length(grep("SHANK3",CNV.ND.Line$GENES))>0 & CNV.ND.Line$SIZE>1000000){
      CNV.ND.Line$CRITERIA_MET=1    
        } 
   
   
   CNV.ND.Line
   
   })))
})))

patho.criteria.met=cnv.patho.criteria[which(cnv.patho.criteria$CRITERIA_MET==1),]

### Deal with smaller nested CNVs 

patho.criteria.met.no.nested=as.data.frame(do.call(rbind,lapply(unique(patho.criteria.met$ID),function(x){
  ### Select row with individual and criteria met

  a=patho.criteria.met[which(patho.criteria.met$ID==x),]
  
  if(nrow(a)==1){
    a
  } else if(nrow(a)>1){
    
    ### Check nested TAR / 1q21, remove TAR from the results row
    if((length(grep("TAR",a$V1))+length(grep("1q21",a$V1)))==2){
      b=a[-grep("TAR",a$V1),]
      b
    }
    
    ### Check nested 15q11.2 in PWS/AS 
    
    if((length(grep("15q",a$V1))+length(grep("PWS",a$V1)))==2){
      b=a[-grep("15q",a$V1),]
      b
    }
    
    ### Check nested 16p11.2 distal within 
    
        if((length(grep("distal",a$V1))+length(grep("16p11.2 del",a$V1)))==2){
      b=a[-grep("distal",a$V1),]
      b
        }
    
  }
})))

###

patho.criteria.met=patho.criteria.met.no.nested
patho.criteria.met=patho.criteria.met[order(as.numeric(as.character(patho.criteria.met[,9])),as.numeric(as.character(patho.criteria.met[,10]))),]
cnv.patho=as.data.frame(table(patho.criteria.met$V1))
cnv.patho=unique(merge(patho.criteria.met[,c(12,9,10)],cnv.patho,by.x="V1",by.y="Var1",sort=F)[,c(1,4)])
cnv.patho$DATASET=DATASET
names(cnv.patho)=c("Neurodevelopmental CNV","N","DATASET")
write.table(cnv.patho.criteria,file=paste(outpath,DATASET,".Patho.Annotated.txt",sep=""),col.names=T,row.names=F,quote=F,sep="\t")
```
### CNV Filtering Parameters

CNVs are subject to specific QC parameters:   
**Number of probes tagging CNVs**: CNVs containing fewer than 20 probes are removed.  
**Confidence score**: reflects the confidence that the CNV is not a false positive finding. CNVs with CONF<10 are removed. 
**Density per probe < 20kb**: This ensures adequate coverage of the CNV by probes.  
`r nrow(CNV.Calls.Exclude)` CNVs failed QC (either <20 probes, confidence score < 10 or density per probe < 20kb).  

# Neurodevelopmental CNV Calls
To be called as genuine ND CNVs, called CNVs undergo further checks that are specific to each locus. These checks could include one or more of the following:  

**1)** The CNV is required to hit specific genes at a locus  
**2)** Exons of a gene may be required to be hit (e.g NRXN1)  
**3)** The CNV may be required to be a certain length (ie greater than 1MB)  
**4)** The CNV may be required to be longer than a certain percentage of the "critical region"  
  
The locus specific criteria for calling neurodevelopmental CNVs were taken from [Kendall el al, 2019](https://jamanetwork.com/journals/jamapsychiatry/fullarticle/2730725). These criteria can be found at the end of this report. 
`r nrow(patho.criteria.met)` CNVs passed specific criteria meaning they may be ND CNVs: 

```{r pathocnv}
cnv.patho
```

### CNV Plots  
For CNVs that pass critical criteria it is essential to manually inspect the B-Allele Frequency and Log R Ratio distributions to add certainty to the CNV calls. Despite merging CNVs that are close together, in some instances, the CNV may in reality be longer than the coordinates called by PennCNV. The plots below show the B-Allele Frequency and Log R Ratio distribution for each individual in the table above. The green dotted lines are the critical coordinates for that ND CNV. Please check these plots carefully to assess whether the called CNV boundries look valid and whether the trace plots look clean.      

**CNV deletions** are characterised by a lack of heterozygous calls (ie no B-Allele frequencies at 0.5). Log R Ratio values < 0 for the length of the deletion.  
**CNV duplications** are characterised by 4 bands on the B-Allele frequency plots (0, 0.25-0.5, 0.5-0.75 and 1). Log R Ratio values are always > 0 for the length of the duplication.  

```{r Pathogenic CNV trace plots}
invisible(lapply(unique(patho.criteria.met$ID),function(x){
patho.id=patho.criteria.met[which(patho.criteria.met$ID==x),]
patho.cnv.lower.coords=as.numeric(as.character(patho.id[,10]))-(patho.id[1,3]*1.3)
patho.cnv.upper.coords=as.numeric(as.character(patho.id[,11]))+(patho.id[1,3]*1.3)
patho.cnv.chr=as.numeric(as.character(patho.id[,9]))
cnv.raw=as.data.frame(fread(input=paste(outpath,x,sep=""),header=T,sep="\t"))
cnv.raw.params=cnv.raw[which(cnv.raw$Position>=patho.cnv.lower.coords & cnv.raw$Position<=patho.cnv.upper.coords & cnv.raw$Chr==patho.cnv.chr),]
cnv.raw.params$GROUP=c("Probes outside called CNV")
cnv.raw.params[which(cnv.raw.params$Position>=as.numeric(as.character(patho.id[,10])) & cnv.raw.params$Position<=as.numeric(as.character(patho.id[,11])) & cnv.raw.params$Chr==patho.cnv.chr),]$GROUP=c("Probes within individually called CNV")
cnv.raw.params$GROUP <- factor(cnv.raw.params$GROUP, levels = c("Probes within individually called CNV","Probes outside called CNV"))


baf.col=grep("Allele",names(cnv.raw.params))
logr.col=grep("Log",names(cnv.raw.params))

baf=ggplot(data=cnv.raw.params,aes(x=cnv.raw.params[,3],y=cnv.raw.params[,baf.col],color=as.factor(GROUP))) +
    geom_point(shape=1) +
    xlab("Base Position")+
    ylab("B-Allele Frequency")+
    ggtitle(paste(x,patho.id$V1,sep=" "))+
    geom_hline(yintercept=0.5, linetype="dashed", color = "Black")+
    geom_vline(xintercept=patho.id[,14], linetype="dashed", color = "Green")+
    geom_vline(xintercept=patho.id[,15], linetype="dashed", color = "Green")+
    ylim(c(0,1))+
    labs(color="CNV Probe Legend")



lrr=ggplot(data=cnv.raw.params,aes(x=cnv.raw.params[,3],y=cnv.raw.params[,logr.col],color=as.factor(GROUP))) +
    geom_point(shape=1) +
    xlab("Base Position")+
  ylab("Log R Ratio")+
    ylim(c(-1,1))+
    geom_hline(yintercept=0, linetype="dashed", color = "Black")+
    geom_vline(xintercept=patho.id[,14], linetype="dashed", color = "Green")+
    geom_vline(xintercept=patho.id[,15], linetype="dashed", color = "Green")+
    labs(color="CNV Probe Legend")



grid.arrange(arrangeGrob(baf,lrr,ncol=1))

png(gsub(" ","_",paste(outpath,"NeuroDevelopmentalPlots/",x,".",patho.id$V1,".png",sep="")), width = 4, height = 4, units = 'in', res = 300)
grid.arrange(arrangeGrob(baf,lrr,ncol=1))
dev.off()
}))
```

## Neurodevelopmental CNV Critical Region Parameters 

In some situations, there may be a CNV that does not pass the critial region parameters, for example it may hit 49% of the critial region rather than 50%. **In this instance the CNV will not be called as being a ND CNV.** You may still wish to call this as a ND CNV. All CNVs that at known ND loci (regardless of whether they meet critical region parameters) are written to the file `r paste(DATASET,".Patho.Annotated.txt",sep="")`. You should always check this file to ensure no borderline ND CNVs are missed. In this file, additional information about each CNV including genes that are hit by the CNV, and the proportion of the criticial region hit by the CNV are included.

Each ND CNV has specific criteria that should be met. The following sections are broken up by chromosome, giving both the critical region of the CNV and the criteria that must be met. Please note these are criteria for all ND CNVs and not just those called in this sample. 

### Chromosome 1
1p36 del/dup: Size	>50% of critical region, affecting GABRD  
TAR	del/dup: Size	>50% of critical region  
1q21.1 del/dup: Size >50% of critical region  
```{r cnvcoords1}
names(cnv.neurodev)=c("LOCUS","CHR","START","END","CLASS")
options(max.print=1000)
cnv.neurodev[which(cnv.neurodev$CHR==1),]
```


### Chromosome 2
NRXN1 del: Must include complete deletion of at least one exon  
2q11.2 del: (LMAN2L,ARID5A): Size >50%	of critical	region,	affecting	both LMAN2L and	ARID5A  
2q13 del/dup: Size	>50%	of	critical	region  
2q37 del (HDAC4): Size	>50%	of	critical	region,	affecting	HDAC4  
```{r cnvcoords2}
names(cnv.neurodev)=c("LOCUS","CHR","START","END","CLASS")
options(max.print=1000)
cnv.neurodev[which(cnv.neurodev$CHR==2),]
```

### Chromosome 3        
3q29 del: Size	>50%	of	critical	region  
```{r cnvcoords3}
names(cnv.neurodev)=c("LOCUS","CHR","START","END","CLASS")
options(max.print=1000)
cnv.neurodev[which(cnv.neurodev$CHR==3),]
```
### Chromosome 4
Wolf‐Hirschhorn	del/dup: Size >50% of critical region  		
```{r cnvcoords4}
names(cnv.neurodev)=c("LOCUS","CHR","START","END","CLASS")
options(max.print=1000)
cnv.neurodev[which(cnv.neurodev$CHR==4),]
```
### Chromosome 5
Sotos syndrome del: Size >50% of critical region  
```{r cnvcoords5}
names(cnv.neurodev)=c("LOCUS","CHR","START","END","CLASS")
options(max.print=1000)
cnv.neurodev[which(cnv.neurodev$CHR==5),]
```
### Chromosome 7
Williams-Beuren Syndrome del/dup:	Size >50% of critical region  		
```{r cnvcoords7}
names(cnv.neurodev)=c("LOCUS","CHR","START","END","CLASS")
options(max.print=1000)
cnv.neurodev[which(cnv.neurodev$CHR==7),]
```
### Chromosome 8   
8p23.1 del/dup: At least 1MB of critical region		
```{r cnvcoords8}
names(cnv.neurodev)=c("LOCUS","CHR","START","END","CLASS")
options(max.print=1000)
cnv.neurodev[which(cnv.neurodev$CHR==8),]
```
### Chromosome 9
9q34 dup (EHMT1): CNV must be at	least	1MB	CNVs,	including	EHMT1  
```{r cnvcoords9}
names(cnv.neurodev)=c("LOCUS","CHR","START","END","CLASS")
options(max.print=1000)
cnv.neurodev[which(cnv.neurodev$CHR==9),]
```
### Chromosome 10
10q23	del: At least 1Mbp,	including	NRG3 and	GRID1  
```{r cnvcoords10}
names(cnv.neurodev)=c("LOCUS","CHR","START","END","CLASS")
options(max.print=1000)
cnv.neurodev[which(cnv.neurodev$CHR==10),]
```
### Chromosome 11   
Potocki‐Shaffer	syndrome del (EXT2) / 11p11.2 dup (EXT2): Size	>50%	of	critical	region,	including	EXT2
```{r cnvcoords11}
names(cnv.neurodev)=c("LOCUS","CHR","START","END","CLASS")
options(max.print=1000)
cnv.neurodev[which(cnv.neurodev$CHR==11),]
```
### Chromosome 15
15q11.2 del/dup BP1-BP2: Size >50% of critical region  
PWS/AS	del/dup: Size = 80% of critical region and must be over 4MB   
15q13.3 del BP4-BP5: Size >50% of critical region  
15q24	del/dup: At least 1Mbp between the	A‐E	intervals  	
15q25	del: At	least	1Mbp	between	the	A‐D	intervals  
```{r cnvcoords15}
names(cnv.neurodev)=c("LOCUS","CHR","START","END","CLASS")
options(max.print=1000)
cnv.neurodev[which(cnv.neurodev$CHR==15),]
```

### Chromosome 16   
16p13.11 del/dup: Size >50%	of critical region  
16p12.1	del: Size >50% of critical region  
16p11.2	distal del/dup: Size >50%	of critical region		
16p11.2	del/dup: Size	>50% of critical region  
```{r cnvcoords16}
names(cnv.neurodev)=c("LOCUS","CHR","START","END","CLASS")
options(max.print=1000)
cnv.neurodev[which(cnv.neurodev$CHR==16),]
```

### Chromosome 17   
17p13.3	del/dup	(YWHAE): Exonic	deletions, whole	gene	duplications  
17p13.3	del/dup	(PAFAH1B1) Exonic	deletions;	whole	gene	duplications  
17p12: >50% of critical region
17q11.2	del/dup	(NF1): Size >50% of critical	region,	affecting	NF1  
Smith‐Magenis/Potocki‐Lupski Syndrome del/dup: Size	>50% of critical region  		
Renal cysts and diabetes syndrome del / 17q12 dup: Size >50%	of critical region  		
17q21.31	del: Size >50% of critical region  		
```{r cnvcoords17}
names(cnv.neurodev)=c("LOCUS","CHR","START","END","CLASS")
options(max.print=1000)
cnv.neurodev[which(cnv.neurodev$CHR==17),]
```

### Chromosome 22   
22q11.2	del/dup: Size >50% of critical region		
22q11.2	distal del/dup: Size >50% of critical region  
SHANK3 del/dup:	 At	least	1Mbp	CNVs,	including	SHANK3 
```{r cnvcoords22}
names(cnv.neurodev)=c("LOCUS","CHR","START","END","CLASS")
options(max.print=1000)
cnv.neurodev[which(cnv.neurodev$CHR==22),]
```
